### PR TITLE
SC: update homi docker compose for kspn and ksen

### PR DESCRIPTION
**Pull request by jeonghonah**
## Proposed changes

- Update to reflect kspn and ksen when creating a homi's docker compose file.

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [ ] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

n/a

## Further comments

n/a

### CODE REVIEW

File: cmd/homi/docker/compose/homi.go [hunk 1]
```diff
@@ -39,19 +39,25 @@ type Homi struct {
 }
 
 func New(ipPrefix string, number int, secret string, addresses, nodeKeys []string,
```
Comment (KimKyungup) [inline cmd/homi/docker/compose/homi.go]: Just minor comment.
How about replacing  `number` -> `numPNs` or remove it? :)

File: cmd/homi/setup/cmd.go [hunk 2]
```diff
@@ -786,6 +810,66 @@ func makeSCNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string)
 	return scn, scnNodeKeys
 }
 
+func makeSPNNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string) {
```
Comment (KimKyungup) [inline cmd/homi/setup/cmd.go]: Can we merge these 3 functions? :)
makeSCN/SPN/SENNodes?

File: cmd/homi/docker/compose/homi.go [hunk 3]
```diff
@@ -165,18 +172,75 @@ func (ist *Homi) init(number int, addresses, nodeKeys []string, genesis, scGenes
...
+			"",
+			spnNodeKeys[i],
+			"",
+			"",
+			32323+number+numPNs+numENs+numSCNs+i,
+			8551+number+numPNs+numENs+numSCNs+i,
+			61001+number+numPNs+numENs+numSCNs+i,
+			ist.EthStats.Host(),
+			// from subnet ip 10
+			fmt.Sprintf("%v.%v", ist.IPPrefix, number+numPNs+numENs+numSCNs+i+10),
```
Comment (KimKyungup) [inline cmd/homi/docker/compose/homi.go]: Just curious. Do you know why `+10`?

File: cmd/homi/docker/compose/homi.go (hunk 1 above)
Comment (jeonghonah) [reply cmd/homi/docker/compose/homi.go]: The `number` is for CN. I'd like to do refactoring at the next PR.

File: cmd/homi/setup/cmd.go (hunk 2 above)
Comment (jeonghonah) [reply cmd/homi/setup/cmd.go]: I'd like to do refactoring at the next PR.

File: cmd/homi/docker/compose/homi.go (hunk 3 above)
Comment (jeonghonah) [reply cmd/homi/docker/compose/homi.go]: I'm not sure. Just to avoid to alloc '0' case?

File: cmd/homi/setup/cmd.go [hunk 4]
```diff
@@ -786,6 +810,66 @@ func makeSCNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string)
...
+	privKeys, nodeKeys, nodeAddrs := istcommon.GenerateKeys(num)
+
+	var p2pPort uint16
+	var proxies []*ValidatorInfo
+	var proxyNodeKeys []string
+	for i := 0; i < num; i++ {
+		if isWorkOnSingleHost {
+			p2pPort = lastIssuedPortNum
+			lastIssuedPortNum++
```
Comment (nogang) [inline cmd/homi/setup/cmd.go]: lastIssuedPortNum is a global variable. Is there a shared resource issue here?

File: cmd/homi/setup/cmd.go [hunk 5]
```diff
@@ -786,6 +810,66 @@ func makeSCNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string)
...
+				p2pPort,
+				discover.NodeTypeUnknown).String(),
+		}
+		proxies = append(proxies, v)
+		proxyNodeKeys = append(proxyNodeKeys, v.Nodekey)
+	}
+	return proxies, proxyNodeKeys
+}
+
+func makeSENNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string) {
```
Comment (nogang) [inline cmd/homi/setup/cmd.go]: makeSPNNodes and makeSENNodes seem to do exactly the same thing. Do I need to repeat the same code?

File: cmd/homi/setup/cmd.go [hunk 6]
```diff
@@ -756,7 +780,7 @@ func makeEndpoints(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string
 	return endpoints, endpointsNodeKeys
 }
 
-func makeSCNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string) {
+func makeSCNNodes(num int, isWorkOnSingleHost bool) ([]*ValidatorInfo, []string) {
```
[the remark this answers is not in the document]
Comment (jeonghonah) [reply cmd/homi/setup/cmd.go]: Fixed.

File: cmd/homi/setup/cmd.go (hunk 5 above)
Comment (jeonghonah) [reply cmd/homi/setup/cmd.go]: Fixed.

File: cmd/homi/setup/cmd.go (hunk 4 above)
Comment (jeonghonah) [reply cmd/homi/setup/cmd.go]: There is no concurrency issue here.

File: cmd/homi/setup/cmd.go [hunk 7]
```diff
@@ -461,26 +461,26 @@ func gen(ctx *cli.Context) error {
 		enNodeValidators, enNodeKeys := makeEndpoints(enNum, false)
 		enNodeInfos := filterNodeInfo(enNodeValidators)
```
Comment (nogang) [inline cmd/homi/setup/cmd.go]: is `enNode` mean `endpoint node node`?

File: cmd/homi/setup/cmd.go [hunk 8]
```diff
@@ -509,9 +509,9 @@ func gen(ctx *cli.Context) error {
 			!ctx.BoolT(nografanaFlag.Name),
 			proxyNodeKeys,
 			enNodeKeys,
```
Comment (nogang) [inline cmd/homi/setup/cmd.go]: is `enNode` mean `endpoint node node`?

File: cmd/homi/setup/cmd.go (hunk 7 above)
Comment (jeonghonah) [reply cmd/homi/setup/cmd.go]: Fixed.

File: cmd/homi/setup/cmd.go (hunk 8 above)
Comment (jeonghonah) [reply cmd/homi/setup/cmd.go]: Fixed.

**Pull request merged**